jcl/src/java.base/share/classes/java/security/AccessControlContext.java

```diff
@@ -351,6 +350,36 @@ public AccessControlContext(ProtectionDomain[] fromContext) {
 	this.containPrivilegedContext = true;
 }
 
+AccessControlContext(ProtectionDomain[] pdArray, AccessControlContext parent, AccessControlContext acc, int authorizeState) {
+	super();
+	switch (authorizeState) {
+	default:
+		// authorizeState can't be STATE_UNKNOWN, callerPD always is NULL
+		throw new IllegalArgumentException();
+	case STATE_AUTHORIZED:
+		if (null != acc) {
+			// when parent combiner is not null, use parent combiner to combine the current context
+			if (parent.getCombiner() != null) {
+				this.context = parent.getCombiner().combine(pdArray, acc.context);
+				this.domainCombiner = parent.getCombiner();
+			} else {
+				this.context = combinePDObjs(pdArray, acc.context);
+				this.domainCombiner = acc.domainCombiner;
+			}
+		} else {
+			this.domainCombiner = parent.domainCombiner;
+			this.context = pdArray;
+			this.nextStackAcc = parent;
+		}
+		break;
+	case STATE_NOT_AUTHORIZED:
+		break;
+	}
+	this.doPrivilegedAcc = acc;
+	this.authorizeState = authorizeState;
+	this.containPrivilegedContext = true;
```

Comment:
Can `AccessControlContext(AccessControlContext acc, ProtectionDomain[] context, int authorizeState)` invoke this constructor to avoid code duplication?

Reply:
Dont think so, `AccessControlContext(AccessControlContext acc, ProtectionDomain[] context, int authorizeState)` has other checks than the added one there.

Reply:
Should be in a new constructor.